plugins/image-prioritizer/class-image-prioritizer-background-image-styled-tag-visitor.php

```diff
@@ -71,6 +95,57 @@ public function __invoke( OD_Tag_Visitor_Context $context ): bool {
 			);
 		}
 
+		$this->lazy_load_bg_images( $context );
+
 		return true;
 	}
+
+	/**
+	 * Optimizes an element with a background image based on whether it is displayed in any initial viewport.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param OD_Tag_Visitor_Context $context Tag visitor context, with the cursor currently at block with a background image.
+	 */
+	private function lazy_load_bg_images( OD_Tag_Visitor_Context $context ): void {
+		$processor = $context->processor;
+
+		// Lazy-loading can only be done once there are URL Metrics collected for both mobile and desktop.
+		if (
+			$context->url_metric_group_collection->get_first_group()->count() === 0
+			||
+			$context->url_metric_group_collection->get_last_group()->count() === 0
+		) {
+			return;
+		}
+
+		$xpath = $processor->get_xpath();
+
+		$in_any_initial_viewport = $context->url_metric_group_collection->is_element_positioned_in_any_initial_viewport( $xpath );
+
+		// If the element is in the initial viewport, do not lazy load its background image.
+		if ( true === $in_any_initial_viewport || null === $in_any_initial_viewport ) {
+			return;
+		}
+
+		$processor->add_class( self::LAZY_BG_IMAGE_CLASS_NAME );
+
+		if ( ! $this->added_lazy_styles ) {
+			$processor->append_head_html(
+				'<style>
+					@media (scripting: enabled) {
+						.has-background.od-lazy-bg-image {
```

Comment:
Where is this `has-background` class coming from? I don't think we should depend on that being there as any element that has the inline style attribute with the background image is eligible for this lazy loading.

Reply:
In Gutenberg, the `has-background` class is typically added to blocks when a background image or color is applied. However, you're right that this might not be consistent, especially with custom blocks or themes.

Reply:
Done in ce351c2.